### PR TITLE
Updates for Orders refactor

### PR DESCRIPTION
**Pull request by birkholz**
Changes:

* Adds a new class `GenericRemoteForeignKey` documented [here](https://github.com/ZeroCater/zc_common/blob/69122d1d7b149ac12bb9d590335c27f846cab066/zc_common/remote_resource/README.md#genericremoteforeignkey-models)
* Fixes several issues with the test mixins
* Adds `IsStaffPermission`

### CODE REVIEW

**Comment by birkholz** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;hgxgfRsI&#x2F;1168-create-genericremoteforeignkey-field-in-zc-common"> Create GenericRemoteForeignKey field in zc_common</a>

File: zc_common/remote_resource/request.py [hunk 1]
```diff
@@ -80,19 +80,30 @@ def add_items_from_data(self):
...
-    for route in routes.iterkeys():
-        if 'resource_type' in routes[route] and routes[route]['resource_type'] == resource_type:
-            if isinstance(pk, (list, set)):
-                expanded = '{}?filter[id__in]={}'.format(expand(route, {}), ','.join([str(x) for x in pk]))
-            else:
-                expanded = expand(route, {'id': pk})
-            return '{0}{1}'.format(routes[route]['domain'], expanded)
-
-    raise RouteNotFoundException('No route for resource_type: "{0}"'.format(resource_type))
+class RouteManager(object):
```
Comment (tnsengimana) [inline zc_common/remote_resource/request.py]: @birkholz What is the practical reason of using a RouteManager?

1. Initially, changing the gateway routes required almost no changes to existing services. But now, each time gateway routes are updated services must be restarted if they want to use the changes introduced.
2. We have a plan to change the way we register routes to allow services to rather register their routes with gateway automatically at boot time (instead of hard coding them in routes.js). We would not be able to take advante of that gateway change with this RouteManager (obviously because of the caching).
3. When running unit tests, I am restricted to populate the cache with initial routes. I don't see how that is any helpful.
4. A much less-restrictive approach would be to cache the first initial request to the gateway (and not initiating the request before the service even attempt to make one). At least this way unit tests will pass without additional changes. Even then, refreshing the cache maybe after a few minutes would be a better way than hitting the gateway once and never do so again.

Comment (birkholz) [inline zc_common/remote_resource/request.py]: Before this change, any time you called `get_route_from_fk`, you would make a request to the Gateway. This meant that if your API response has a multiple remote FKs (OrderItem has 2), you would make multiple redundant requests for each field in the response. Even just 1 request is unacceptable within the request scope, because it harshly affects response times. That means the only acceptable solution would be to cache the route list outside of the request cycle. `RouteManager` is the simplest solution to the problem without building a complex solution involving memcache or redis and a worker process to refresh the route list on a timer. Our route lists are not currently changing all that often, and services are not dynamically reacting to new routes, so this is not a problem we need to solve.

Now to address your points:

1. Anytime a service is modified to access a new route, it's being restarted anyways as part of the deployment process. This doesn't change anything. You just have to make sure the gateway is updated first.
2. Even if services register their routes automatically, that's only one part of the problem. The other part is somehow allowing services to stay up to date with the route list. A solution to this problem is still a long ways away, and there is no proposed solution for the 2nd half of the problem yet. In the meantime, `RouteManager` is an acceptable stopgap.
3. You're not modifying the gateway's route list in the middle of tests, so this changes nothing. In fact, this will speed up tests over the previous version considerably for the reasons mentioned above.
4. Caching at request time is unacceptable because it affect's that request's response time. Again, like I said in the previous point, unit tests are unaffected. Refreshing on a timer is impossible without a worker process and redis/memcached, and as I said above, that's way more complicated than we need right now.

Comment (tnsengimana) [inline zc_common/remote_resource/request.py]: @birkholz I understand the performance improvement you want to make by this `RouteManager`. However, the implementation is a bit inflexible. How am supposed to conveniently populate the cache?

You implementation assumes the `GATEWAY_ROOT_PATH` is always accessible which is false when running unit tests. However, by caching the first request, at test which hits `GATEWAY_ROOT_PATH` will then get its response cached and thus have unit tests running without blocking. I can't test my changes now because of this problem.

Comment (tnsengimana) [inline zc_common/remote_resource/request.py]: Let me know once you get this issue resolved to move ahead with this branch https://github.com/ZeroCater/mp-slots-and-orders/pull/107

Comment (birkholz) [inline zc_common/remote_resource/request.py]: @tnsengimana As I've already stated multiple times, there is no way to update the route list across multiple processes without a worker process and a store that the web processes can read from, and that is an extremely expensive and complex solution that is wholly unacceptable right now.

> You implementation assumes the GATEWAY_ROOT_PATH is always accessible which is false when running unit tests. However, by caching the first request, the test which hits GATEWAY_ROOT_PATH will then get its response cached and thus have unit tests running without blocking.

This doesn't make any sense. You're saying `GATEWAY_ROOT_PATH` is not accessible from tests, so the `RouteManager` won't work, so we should just cache the first request, which would require the gateway to be available. It's either available, or its not.

However, you do bring up a good point. During testing, we shouldn't make a request to the gateway, because then we'd be integrating against it. Instead, `get_route_from_fk` should do something locally in the case of `DEBUG = True`. I'll make that change and push it up, and that should fix your testing issue.

Comment (birkholz) [inline zc_common/remote_resource/request.py]: @tnsengimana Why are you developing against this branch?

Comment (tpilara) [inline zc_common/remote_resource/request.py]: Another consideration here is that with this change the startup order of our orchestration script matters. If the gateway starts up after any service using the newest version of zc_common, the service will crash and not start up.

Reordering the startup order of services may help resolve this, but it's not guaranteed. We could also add a short sleep/retry in the zc_common script (which would add a little more resilience, even if prod if the gateway is momentarily unreachable), but this is not ideal either.

File: zc_common/remote_resource/relations.py [hunk 2]
```diff
@@ -6,23 +6,19 @@
 from rest_framework_json_api.relations import ResourceRelatedField
 
 from zc_common.remote_resource.models import RemoteResource
+from zc_common.remote_resource.request import get_route_from_fk
 
 
 class RemoteResourceField(ResourceRelatedField):
 
-    def __init__(self, related_resource_path=None, **kwargs):
```
Comment (hamedahmadi) [inline zc_common/remote_resource/relations.py]: I would suggest to not change the behavior of this Model for now. Lets put back all the changes.

File: zc_common/remote_resource/tests.py [hunk 3]
```diff
@@ -50,8 +51,14 @@ def validate_instance_in_data(self, data, instance, attributes, relationship_key
             if hasattr(instance_attribute, '__call__'):
                 instance_attribute = instance_attribute()
 
-            self.assertEqual(
```
Comment (hamedahmadi) [inline zc_common/remote_resource/tests.py]: Why is this removed?

File: zc_common/remote_resource/tests.py [hunk 4]
```diff
@@ -127,24 +134,10 @@ def success_response_structure_test(self, response, status, relationship_keys=No
                     self.assertTrue(
                         all(key in relationship for key in ['data', 'links']))
 
-                    for relationship_data in self.convert_to_list(relationship['data']):
```
Comment (hamedahmadi) [inline zc_common/remote_resource/tests.py]: Why are we removing these parts of the test? they seem to be testing very critical parts of the response. I would agree that the attributes that this method uses could be better documented, but lets refrain from removing the tests.

File: zc_common/remote_resource/test_mixins.py [hunk 5]
```diff
@@ -386,7 +391,7 @@ def test_resource(self):
         self.assertIsNotNone(deleted_resource[0].deleted_at)
 
     def test_delete__404_for_nonexistant_resource(self):
-        url = reverse(self.resource_view_name, args=(self.resource.id + 10000,))
```
Comment (hamedahmadi) [inline zc_common/remote_resource/test_mixins.py]: Why are we changing these values?

File: zc_common/remote_resource/models.py [hunk 6]
```diff
@@ -51,3 +52,95 @@ def deconstruct(self):
         del kwargs['max_length']
 
         return name, path, args, kwargs
+
+
+class GenericRemoteForeignKey(object):
```
Comment (hamedahmadi) [inline zc_common/remote_resource/models.py]: This has a lot of logic that are not easy to verify, how have we tested this? are we planning to write tests for it? How confident are we that this is stable and reliable?

File: zc_common/jwt_auth/permissions.py [hunk 7]
```diff
@@ -63,3 +63,8 @@ def has_update_permission(self, request, view):
 class EventViewPermission(BasePermission):
     def has_create_permission(self, request, view):
         return is_service(request)
+
+
+class IsStaffPermission(permissions.BasePermission):
```
Comment (hamedahmadi) [inline zc_common/jwt_auth/permissions.py]: Why are we adding this permission class here? if this is a good place for it, why we don't have other permission classes like User or service here as well?

File: zc_common/remote_resource/test_mixins.py (hunk 5 above)
Comment (tpilara) [inline zc_common/remote_resource/test_mixins.py]: How it was previously written works when the resource has an integer as its primary key, but if you have something with a non-integer such as a UUID, Python will throw an error because it's trying to perform addition on a str and an int. Since we're trying to test for a 404/nonexistent resource, if we replace the entire calculation with an id we know will not exist as an integer and pass it via args to reverse here, Django will correctly cast it to a string or an integer depending on the field type.

File: zc_common/remote_resource/models.py (hunk 6 above)
Comment (birkholz) [inline zc_common/remote_resource/models.py]: This class is an almost exact copy of [Django's GenericForeignKey](https://github.com/django/django/blob/master/django/contrib/contenttypes/fields.py#L21). The testing in slots and orders should be sufficient.

**Comment by birkholz** [comment]
The `RouteManager` and other changes to `get_route_from_fk` have been removed, due to some issues that were not easy to solve.

File: zc_common/remote_resource/test_mixins.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 import json
 
+import datetime
+import dateutil.parser
```
Comment (hamedahmadi) [inline zc_common/remote_resource/test_mixins.py]: having this means we have to add this to every single service, just to be able to run one small part of a test.

Comment (birkholz) [inline zc_common/remote_resource/test_mixins.py]: Unfortunately, the parser is required for the test mixin to work properly. You'd only need to install the requirement when running tests, though, so you don't necessarily have to add it to the `requirements.txt`, you could just add it to the `circle.yml` file. However, adding `dateutil` to `requirements.txt` won't affect prod at all.

This is one of the downsides of the approach we chose to take with zc_common, which is that we don't force installation of zc_common's dependencies. So you can import zc_common and only use the timezone code and not have to install DRF/json api, for example.

Ideally, we'd split out all the DRF related code into its own repo/project and force the dependencies. So when you install that project, it'd auto-install its dependencies. For now, I'd rather not go through the trouble due to the number of changes being added frequently. I think that's the step we should take sometime in the next few months. Essentially, split up zc_common into:

1. `zc_microservices`: a package containing all our microservice shared code (JSON API, remote resource, etc)
2. `zc_time`: a package containing all our utilities for handling dates and times

**Comment by hamedahmadi** [comment]
[Build links for remote resources automatically (proper generic FK support)](https://trello.com/c/YaWiKykz/1336-build-links-for-remote-resources-automatically-proper-generic-fk-support)

File: zc_common/remote_resource/tests.py (hunk 3 above)
Comment (birkholz) [inline zc_common/remote_resource/tests.py]: It was replaced with a better check that works for datetimes.

File: zc_common/remote_resource/tests.py (hunk 4 above)
Comment (birkholz) [inline zc_common/remote_resource/tests.py]: Added back (https://github.com/ZeroCater/zc_common/pull/62/commits/d5bfd3210e769491ec8cd23b09974a718c3db181)
